sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/BeamFnDataGrpcMultiplexer.java

```diff
@@ -0,0 +1,132 @@
+ *
+ * <p>Multiplexes data for inbound consumers based upon their individual
+ * {@link org.apache.beam.fn.v1.BeamFnApi.Target}s.
+ *
+ * <p>TODO: Add support for multiplexing over multiple outbound observers by stickying
+ * the output location with a specific outbound observer.
+ */
+public class BeamFnDataGrpcMultiplexer {
+  private static final Logger LOGGER = LoggerFactory.getLogger(BeamFnDataGrpcMultiplexer.class);
+  private final BeamFnApi.ApiServiceDescriptor apiServiceDescriptor;
+  private final StreamObserver<BeamFnApi.Elements> inboundObserver;
+  private final StreamObserver<BeamFnApi.Elements> outboundObserver;
+  @VisibleForTesting
+  final ConcurrentMap<KV<Long, BeamFnApi.Target>,
+                              CompletableFuture<Consumer<BeamFnApi.Elements.Data>>> consumers;
+
+  public BeamFnDataGrpcMultiplexer(
+      BeamFnApi.ApiServiceDescriptor apiServiceDescriptor,
+      Function<StreamObserver<BeamFnApi.Elements>,
+               StreamObserver<BeamFnApi.Elements>> outboundObserverFactory) {
+    this.apiServiceDescriptor = apiServiceDescriptor;
+    this.consumers = new ConcurrentHashMap<>();
+    this.inboundObserver = new InboundObserver();
+    this.outboundObserver = outboundObserverFactory.apply(inboundObserver);
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("apiServiceDescriptor", apiServiceDescriptor)
+        .add("consumers", consumers)
+        .toString();
+  }
+
+  public StreamObserver<BeamFnApi.Elements> getInboundObserver() {
+    return inboundObserver;
+  }
+
+  public StreamObserver<BeamFnApi.Elements> getOutboundObserver() {
+    return outboundObserver;
+  }
+
+  public CompletableFuture<Consumer<BeamFnApi.Elements.Data>> futureForKey(
+      KV<Long, BeamFnApi.Target> key) {
+    return consumers.computeIfAbsent(
+        key,
+        (KV<Long, BeamFnApi.Target> providedKey) -> new CompletableFuture<>());
+  }
+
+  /**
+   * A multiplexing {@link StreamObserver} that selects the inbound {@link Consumer} to pass
+   * the elements to.
+   *
+   * <p>The inbound observer blocks until the {@link Consumer} is bound allowing for the
+   * sending harness to initiate transmitting data without needing for the receiving harness to
+   * signal that it is ready to consume that data.
+   */
+  private final class InboundObserver implements StreamObserver<BeamFnApi.Elements> {
+    @Override
+    public void onNext(BeamFnApi.Elements value) {
+      for (BeamFnApi.Elements.Data data : value.getDataList()) {
+        try {
+          KV<Long, BeamFnApi.Target> key =
+              KV.of(data.getInstructionReference(), data.getTarget());
+          futureForKey(key).get().accept(data);
+          if (data.getData().isEmpty()) {
+            consumers.remove(key);
+          }
+        } catch (ExecutionException | InterruptedException e) {
+          LOGGER.error(
```

Comment:
Added TODO, since this really needs to be wired up to fail the bundle processing of all impacted bundles.